dali/operators/reader/loader/numpy_loader.cc

```diff
@@ -0,0 +1,178 @@
+  else if (format == "i4")
+    return TypeInfo::Create<int32_t>();
+  else if (format == "f4")
+    return TypeInfo::Create<float>();
+  else if (format == "f8")
+    return TypeInfo::Create<double>();
+  else
+    return TypeInfo();
+}
+
+std::unique_ptr<FileStream> NumpyLoader::ParseHeader(std::unique_ptr<FileStream> file,
+                                                     NumpyParseTarget& target) {
+  // check if the file is actually a numpy file
+  uint8_t* token = new uint8_t[10];
+  int64_t nread = file->Read(token, 10);
+  DALI_ENFORCE(nread == 10, "Can not read header.");
+
+  // check if heqder is too short
+  char* chartoken = reinterpret_cast<char*>(token);
+  DALI_ENFORCE(std::string(chartoken).find("NUMPY") != std::string::npos,
+               "File is not a numpy file.");
+
+  // extract header length
+  uint16_t header_len = 0;
+  memcpy(&header_len, &token[8], 2);
+  delete [] token;
+
+  // read header: the offset is a magic number
+  int64 offset = (6+1+1+2);
+  token = new uint8_t[header_len];
+  chartoken = reinterpret_cast<char*>(token);
+  file->Seek(offset);
+  nread = file->Read(token, header_len);
+  DALI_ENFORCE(nread == header_len, "Can not read header.");
+  DALI_ENFORCE(std::string(chartoken).find("{") != std::string::npos, "Header is corrupted.");
+  std::string header(chartoken);
+  delete [] token;
+  offset += header_len;
+
+  // prepare file for later reads
+  file->Seek(offset);
+
+  // extract dictionary info from header
+  std::smatch header_match;
+  DALI_ENFORCE(std::regex_search(header, header_match, header_regex_),
+               "Can not parse header.");
+
+  // now extract header information
+  // type
+  std::string typestring = header_match[1].str();
+
+  // < means LE, | means N/A, = means native. In all those cases, we can read
+  bool little_endian =
+    (typestring[0] == '<' || typestring[0] == '|' || typestring[0] == '=' ? true : false);
+  DALI_ENFORCE(little_endian,
+    "Big Endian files are not supported.");
+
+  std::string tid = typestring.substr(1);
+  // get type in a safe way
+  target.type_info = TypeFromNumpyStr(tid);
+
+  // check for data order
+  if (header_match[2].str() == "False")
+    target.fortran_order = false;
+  else
+    target.fortran_order = true;
+
+  // set sizes
+  std::string shapestring = header_match[3].str();
+  std::regex shape_regex{R"(,+)"};  // split on comma
+  std::sregex_token_iterator it{shapestring.begin(), shapestring.end(), shape_regex, -1};
+  std::vector<std::string> shapevec{it, {}};
+
+  // if shapevec size is 1 and shapevec[0] is the empty string,
+  // it is a scalar/singleton (denoted as ()) and needs to be set to one:
+  if ( (shapevec.size() == 1) && (shapevec[0] == "") ) shapevec[0] = "1";
+
+  // determine shapes
+  int64_t shapesize = shapevec.size();
+  target.shape.resize(shapesize);
+  // cheapest thing to do is to define the tensor in an reversed way
+  if (target.fortran_order) {
+    for (uint16_t i = 0; i < shapesize; ++i)
```

Comment:
Why uint16_t ?

Reply:
Do you have a preferred datatype for it? Like size_t seems overkill.

Reply:
I don't think we need to fight for every byte of program memory, we are not in 80s.
@mzient, what do you think ?

Reply:
changed to size_t